env/io_posix.cc

```diff
@@ -736,79 +749,111 @@ IOStatus PosixRandomAccessFile::MultiRead(FSReadRequest* reqs, size_t num_reqs,
       io_uring_sqe_set_data(sqe, rep_to_submit);
       wrap_cache.emplace(rep_to_submit);
     }
-    incomplete_rq_list.clear();
-
-    ssize_t ret =
-        io_uring_submit_and_wait(iu, static_cast<unsigned int>(this_reqs));
-    TEST_SYNC_POINT_CALLBACK(
-        "PosixRandomAccessFile::MultiRead:io_uring_submit_and_wait:return1",
-        &ret);
-    TEST_SYNC_POINT_CALLBACK(
-        "PosixRandomAccessFile::MultiRead:io_uring_submit_and_wait:return2",
-        iu);
-
-    if (static_cast<size_t>(ret) != this_reqs) {
-      fprintf(stderr, "ret = %ld this_reqs: %ld\n", (long)ret, (long)this_reqs);
-      // If error happens and we submitted fewer than expected, it is an
-      // exception case and we don't retry here. We should still consume
-      // what is is submitted in the ring.
-      for (ssize_t i = 0; i < ret; i++) {
-        struct io_uring_cqe* cqe = nullptr;
-        io_uring_wait_cqe(iu, &cqe);
-        if (cqe != nullptr) {
-          io_uring_cqe_seen(iu, cqe);
+    resubmit_rq_list.clear();
+
+    struct io_uring_cqe* cqe = nullptr;
+    unsigned head;
+    if (reqs_to_submit > 0) {
+      size_t reqs_submitted = 0;
+      ssize_t err = 0;
+      bool drain_completion_queue = false;
+      do {
+        ssize_t ret = io_uring_submit_and_wait(
+            iu, static_cast<unsigned int>(reqs_to_submit - reqs_submitted));
+        TEST_SYNC_POINT_CALLBACK(
+            "PosixRandomAccessFile::MultiRead:io_uring_submit_and_wait:return1",
+            &ret);
+        TEST_SYNC_POINT_CALLBACK(
+            "PosixRandomAccessFile::MultiRead:io_uring_submit_and_wait:return2",
+            iu);
+        if (ret < 0) {
+          if (-EINTR == ret) {
+            // Submission failed due to interrupted system call. Try again.
+            continue;
+          }
+          if (-EAGAIN == ret || -ENOMEM == ret) {
+            fprintf(stderr,
+                    "PosixRandomAccessFile::MultiRead: "
+                    "io_uring_submit_and_wait returned transient error = %zd\n",
+                    ret);
+            // Best effort to reclaim resources in terse condition.
+            drain_completion_queue = true;
+          } else {
+            fprintf(stderr,
+                    "PosixRandomAccessFile::MultiRead: "
+                    "io_uring_submit_and_wait returned terminal error: %zd\n",
+                    ret);
+            err = ret;
+          }
+          break;
+        }
+        if (0 == ret) {
+          // This scenario is unexpected for any modern kernel!
+          fprintf(stderr,
+                  "PosixRandomAccessFile::MultiRead: "
+                  "io_uring_submit_and_wait returned 0 submissions!\n");
+          break;
         }
+        reqs_submitted += static_cast<size_t>(ret);
+      } while (reqs_submitted < reqs_to_submit);
```

Comment:
What is the state of the IO uring if reqs_submitted < reqs_to_submit? I guess we can trust the kernel to know where it left off so it can pick up from there?